### ISSUE TITLE
Swagger `@Schema` annotation on value types does not persist to schema properties

### ISSUE THREAD

**Issue by rmcmk** [SMILEY4/schema-kenerator#27]
`@Schema` annotations on an inline value class should propagate to the property after inlining. 

For example:
```kt
@JvmInline
@Serializable
@Schema(
    format = "hex",
    description = "Represents a hexadecimal string."
)
value class HexString private constructor(val value: String) {
    // ... remainder omitted for brevity
}
```

![image](https://github.com/user-attachments/assets/f6462862-892e-486a-9a3a-0f57f96ad27d)

The `title` and `description` propagate properly as per the `SwaggerSchemaAnnotationStep`

**Comment by NecroKote** [comment]
Feels like a duplicate of #17 again

Edit: nope. nothing to do with it.

**Comment by rmcmk** [comment]
> Feels like a duplicate of #17 again

Happens with reflection and kotlinx-serialization. Believe this should be a simple matter of merging the schema from the base value type with the inline property schema. #17 is still relevant

**Comment by NecroKote** [comment]
It is indeed the case. After some digging I figured that kenerator is missing handling of `value class` completely - it just treat's them as normal classes. 
I hacked something that worked for the reflection case:

In `schema-kenerator-reflection/src/main/kotlin/io/github/smiley4/schemakenerator/reflection/steps/ReflectionTypeProcessingStep.kt`
```kotlin

    private fun parseValueClass(
        type: KType,
        clazz: KClass<*>,
        typeParameters: Map<String, TypeParameterData>,
        typeData: MutableList<BaseTypeData>
    ): WrappedTypeData {
        // check type redirects
        if (typeRedirects.containsKey(type)) {
            return process(typeRedirects[type]!!, typeData)
        }

        // check custom type processors
        if (customProcessors.containsKey(clazz)) {
            return WrappedTypeData(
                typeData = customProcessors[clazz]!!.invoke().also { result ->
                    typeData.removeIf { it.id == result.id }
                    typeData.add(result)
                },
                nullable = false
            )
        }

        // resolve type parameters, i.e. generic types
        val resolvedTypeParameters = parseTypeParameters(type, clazz, typeParameters, typeData)

        val underlyingClazz = clazz.primaryConstructor?.parameters?.firstOrNull() ?: throw IllegalArgumentException("Value class must have a primary constructor with a single parameter")
        val underlyingType = underlyingClazz.type
        val valueClassAnnotations = parseAnnotations(clazz).filter { it.name != JvmInline::class.qualifiedName }

        return parseClass(underlyingType, underlyingType.classifier as KClass<*>, resolvedTypeParameters, typeData).also {
            it.typeData.annotations.addAll(valueClassAnnotations)
            it.typeData.annotations.addAll(underlyingClazz.annotations.map { ann -> parseAnnotation(ann) })
        }
    }

    @Suppress("CyclomaticComplexMethod", "LongMethod")
    private fun parseClass(
        type: KType,
        clazz: KClass<*>,
        typeParameters: Map<String, TypeParameterData>,
        typeData: MutableList<BaseTypeData>
    ): WrappedTypeData {
        // check if value class
        if (clazz.isValue) {
            return parseValueClass(type, clazz, typeParameters, typeData)
        }
        // ... the rest of this method is unchanged
}

    private fun parseAnnotations(property: KProperty<*>, clazz: KClass<*>): List<AnnotationData> {
        return buildList {
            property.returnType.also { propertyType ->
                propertyType.classifier?.let { classifier ->
                    if (classifier is KClass<*> && classifier.isValue) {
                        addAll(
                            unwrapAnnotations(classifier.annotations.toList())
                                .map(::parseAnnotation)
                                .filter { it.name != JvmInline::class.qualifiedName }
                        )
                    }
                }
            }

            addAll(unwrapAnnotations(property.javaField?.annotations?.toList() ?: emptyList()).map { parseAnnotation(it) })
            addAll(unwrapAnnotations(property.annotations).map { parseAnnotation(it) })
            clazz.primaryConstructor?.parameters
                ?.find { it.name == property.name && it.type == property.returnType }?.annotations
                ?.also { annotations -> addAll(unwrapAnnotations(annotations).map { parseAnnotation(it) }) }
        }
```

This hack allows this:
```kotlin
@JvmInline
@Schema(
    description = "Represents a hexadecimal string.",
    format = "hex",
)
value class HexString(val value: String)

data class Sample(
    val hex: HexString,
)
```

to render like this: 
```json
    "components": {
        "schemas": {
            "io.github.smiley4.ktorswaggerui.examples.Sample": {
                "properties": {
                    "hex": {
                        "format": "hex",
                        "description": "Represents a hexadecimal string.",
                        "maxLength": 2147483647,
                        "minLength": 0,
                        "title": "String"
                    }
                },
                "required": [
                    "hex"
                ],
                "title": "Sample"
            }
        },
        "examples": {}
    }
```

**Comment by SMILEY4** [comment]
Hi, thanks for bringing this up.
after some initial testing, think the processing-steps are (almost) fine. Value classes are treated as normal classes but are marked as "inline value classes" which *should* in theory be enough information for the next steps.
https://github.com/SMILEY4/schema-kenerator/blob/28f02706ccdfba235654c33738001ab98588c015/schema-kenerator-reflection/src/main/kotlin/io/github/smiley4/schemakenerator/reflection/steps/ReflectionTypeProcessingStep.kt#L241
https://github.com/SMILEY4/schema-kenerator/blob/28f02706ccdfba235654c33738001ab98588c015/schema-kenerator-serialization/src/main/kotlin/io/github/smiley4/schemakenerator/serialization/steps/KotlinxSerializationTypeProcessingStep.kt#L291

The value classes are then inlined during schema generation:
https://github.com/SMILEY4/schema-kenerator/blob/28f02706ccdfba235654c33738001ab98588c015/schema-kenerator-swagger/src/main/kotlin/io/github/smiley4/schemakenerator/swagger/steps/SwaggerSchemaGenerationStep.kt#L154

I think there are two ""other"" issues here:

1. when inlining value classes during schema generation, the annotations are dropped / not passed down. I think the cleanest solution here would be to copy the annotation of the value class to the property in the constructor in the processing-step (and maybe mark them as originating from the parent class for traceability).

2.  the annotation on the type does still leak into the schema for the property, though i'm pretty sure that isn't supposed to work like this. I would have expected the annotation to be ignored completly. I think i'll have to look into this aswell.

**Comment by SMILEY4** [comment]
Think i found the underlying issue and a fix for it. 
Tldr: the Schema-annotation here is on the type, not the property. So when handling the annotation here  ...
https://github.com/SMILEY4/schema-kenerator/blob/28f02706ccdfba235654c33738001ab98588c015/schema-kenerator-swagger/src/main/kotlin/io/github/smiley4/schemakenerator/swagger/steps/SwaggerSchemaAnnotationStep.kt#L51
... the step first looks for annotation on types, only handles "title", "description" and "hidden" and then at annotations on properties and handles all the rest, resulting in most properties of the schema annotation on the value class getting ignored.

Steps now look at annotations on the property and on the type of the property. I think it makes more sense this way. Hope this properly fixes it without breaking anything else :)

For changes, see: https://github.com/SMILEY4/schema-kenerator/compare/fix-annotation-on-value-types?expand=1

I'll merge it soon and create a new bugfix version.
